File: leetcode/20. Valid Parentheses.cpp
// Problem link: https://leetcode.com/problems/valid-parentheses/description/
// Time: O(n)
// Space: O(n)

class Solution {
public:
    bool isValid(string s) {
        int n = s.size();
        bool flag = true;
        stack<char> st;
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else {
                if (st.empty() || (c == ')' && st.top() != '(') ||
                    (c == '}' && st.top() != '{') ||
                    (c == ']' && st.top() != '[')) {
                    return false;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};

File: neetcode-150/stack/20. Valid Parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        var st = new Stack<char>();
        foreach (char c in s)
        {
            if (c == '(' || c == '[' || c == '{') st.Push(c);
            else
            {
                if (st.Count() == 0) return false;
                char top = st.Peek();
                if ((top == '(' && c == ')') ||
                        (top == '{' && c == '}') ||
                        (top == '[' && c == ']')) st.Pop();
                else return false;
            }
        }
        return st.Count() == 0;
    }
}